scripts/utils/train_helper.py

```diff
@@ -27,7 +30,56 @@ def get_selfplay_agent(args, total_training_timesteps, train_types, eval_types,
     )
 
     selfplay_trainer.train_agents(total_train_timesteps=total_training_timesteps)
-    return selfplay_trainer.get_agents()
+    return selfplay_trainer.get_agents()[0], tc
+
+
+def get_selfplay_agent_trained_w_selfplay_types(args,
+                                                pop_total_training_timesteps:int,
+                                                sp_train_types:list,
+                                                sp_eval_types:list,
+                                                sp_w_sp_total_training_timesteps:int,
+                                                sp_w_sp_train_types:list,
+                                                sp_w_sp_eval_types:list,
+                                                tag:str=None,
+                                                force_training:bool=None) -> OAIAgent:
+    '''
+    Train a SP agent using SP train types. This function will first train a SP agent and then let that
+    agent train with itself and one other unseen teammate (e.g. [SP, SP, SP, SP_H] in a 4-chef layout)
+    '''
+
+    name = 'sp_w_selfplay_types'
+
+    # TODO: what do we do with this loaded agent?
+    agents = load_agents(args, name=name, tag=tag, force_training=force_training)
+
+    sp_agent, teammates_collection = get_selfplay_agent_w_tms_collection(args=args,
+                                                                          total_training_timesteps=pop_total_training_timesteps,
+                                                                          train_types=sp_train_types,
+                                                                          eval_types=sp_eval_types,
+                                                                          force_training=force_training)
+
+    # Generate a new teammates collection using the original SP agent and its teammates collection
+    # TODO: Can we just use generate_TC_for_FCP_w_SP_types here (maybe also renaming the function)? 
```

Comment:
We cannot, because the teammates_collection returned from `get_selfplay_agent_w_tms_collection` only has SELF_PLAY type in it. But if you return the teammates_collection from `get_fcp_agent_w_tms_clction` you can then call  `generate_TC_for_FCP_w_SP_type`. I think one thing you can do is to call `get_fcp_population` to get the proper teammates_collection and then you can write a function which returns a randomly initialized sb3_agent similar to `get_sb3_agent` in `rl.py` and then you can call  `generate_TC_for_FCP_w_SP_types` with the sb3_agent and the teammates_collection.

Reply:
Ok I am a bit confused then. We are only using `TeamType.SELF_PLAY` for the first SP `train_type` used to to generate the first `teammates_collection` correct?

If that's the case, I don't think I can use `get_fcp_population()` because that assumes that `TeamType.SELF_PLAY` is NOT in the `train_type` argument (this [line](https://github.com/HIRO-group/multiHRI/blob/main/scripts/utils/tc_helper.py#L109) essentially defines that `SELF_PLAY` is not allowed when calling `get_fcp_population()`)

Should I be using a different `TeamType` when calling `get_fcp_population()`?

Reply:
Yes, that is correct. You should call `get_fcp_population` with `X_FIRST`, where `X` is gonna be in `SELF_PLAY_X`. For example, if you want to do SP-SP-H, SP-SP-L then you should do this:

- call `get_fcp_population()` with `[TeamType.HIGH_FIRST, TeamType.LOW_FIRST]` to get `teammates_collection`
- call `get_randomly_initialized_sb3_agent()`  to get a randomly initialized `sb3_agent`
- call `generate_TC_for_FCP_w_SP_types()` with `[TeamType.SELF_PLAY_HIGH, TEAMTYPE.SELF_PLAY_LOW]` and `teammates_collection` and `sb3_agent` to get the `final_teammates_collection`
- Finally, call `RLAgentTrainer` with the `sb3_agent` and `final_teammates_collection` (which you currently have in your [code](https://github.com/HIRO-group/multiHRI/blob/00806ded2e34560a182ac6ff8ac12638995d05ed/scripts/utils/train_helper.py#L71))

Reply:
Ok this is making a lot more sense to me now. Thank you for your patience and detail.
I think I fixed it!